File: 2020/interview/tpc/warmup/tpc-warmup-3-how-many-nines.cpp
#include <iostream>
#include <string>
#include <cstring>

using namespace std;

const int MONTHS = 12;
const int DAYS = 31;
int leap_days[MONTHS] = {31, 29, 31, 30, 31, 30, 31, 31, 30, 31, 30, 31};
int norm_days[MONTHS] = {31, 28, 31, 30, 31, 30, 31, 31, 30, 31, 30, 31};
int count_md[MONTHS][DAYS];
long int pre_sum[10000][MONTHS][DAYS];

int count_int(int n)
{
    int ret = 0;
    while (n > 0) ret += (int)(n % 10 == 9), n /= 10;
    return ret;
}

int is_leap(int y) {
    return y % 400 == 0 || y % 4 == 0 && y % 100 > 0;
}

void compute(int days[MONTHS], int day_sum[][DAYS], int day_count)
{
    int month = 0, day = 0, count = 0;
    while (month < MONTHS)
    {
        count_md[month][day] = count_int(month + 1) + count_int(day + 1);
        day += 1;
        if (day >= days[month]) day = 0, month += 1;
    }
}

void pre_compute()
{
    compute(leap_days, count_md, 366);
    int y = 2000, m = 0, d = 0;
    int count_year = 0;
    long int count = 0;
    int* days = leap_days;
    while (y < 10000)
    {
        count += count_year + count_md[m][d];
        pre_sum[y][m][d] = count;
        if (++d >= days[m]) d = 0, m += 1;
        if (m >= 12)
        {
            m = 0, y += 1;
            days = is_leap(y) ? leap_days : norm_days;
            count_year = count_int(y);
        }
    }
}

int main()
{
    int t;
    cin >> t;
    pre_compute();
    for (int i = 0; i < t; i += 1)
    {
        int sy, sm, sd, ey, em, ed;
        cin >> sy >> sm >> sd >> ey >> em >> ed;
        long int ret = pre_sum[ey][em - 1][ed - 1] - pre_sum[sy][sm - 1][sd - 1];
        ret += count_int(sy) + count_int(sm) + count_int(sd);
        printf("%ld\n", ret);
    }
}

File: 2020/interview/tpc/warmup/tpc-warmup-3-how-many-nines.py
def read_ints():
    return list(map(int, input().split(' ')))

def solve():
    t = int(input())
    leap_days = [31, 29, 31, 30, 31, 30, 31, 31, 30, 31, 30, 31]
    norm_days = [31, 28, 31, 30, 31, 30, 31, 31, 30, 31, 30, 31]
    leap = [0] * sum(leap_days)
    norm = [0] * sum(norm_days)
    for i, days in enumerate([leap_days, norm_days]):
        count = 0
        month, day = 0, 0
        for d in range(sum(days)):
            count += sum(c == '9' for c in '{}{}'.format(month + 1, day + 1))
            [leap, norm][i][d] = count
            day += 1
            if day >= days[month]:
                day = 0
                month += 1

    def is_leap(y):
        return y % 400 == 0 or y % 4 == 0 and y % 100 > 0

    def get_day_index(y, m, d):
        days = [norm_days, leap_days][int(is_leap(y))]
        return sum(days[:m]) + d

    def count_year(y):
        count = 0
        while y:
            count += int(y % 10 == 9)
            y //= 10
        return count

    def count_range(y, sm, sd, em, ed):
        count = [norm, leap][int(is_leap(y))]
        st_day = get_day_index(y, sm, sd)
        ed_day = get_day_index(y, em, ed)
        st_count = 0 if st_day == 0 else count[st_day - 1]
        ed_count = count[ed_day]
        return ed_count - st_count + count_year(y) * (ed_day - st_day + 1)

    for i in range(t):
        sy, sm, sd, ey, em, ed = read_ints()
        sm, sd, em, ed = sm - 1, sd - 1, em - 1, ed - 1
        ret = 0
        for y in range(sy + 1, ey):
            cy = count_year(y)
            if is_leap(y): ret += leap[-1] + cy * 366
            else: ret += norm[-1] + cy * 365
        if sy == ey:
            ret += count_range(sy, sm, sd, em, ed)
        else:
            # 开始年份
            ret += count_range(sy, sm, sd, 11, 30)
            ret += count_range(ey, 0, 0, em, ed)
        print(ret)

solve()

'''
4
2017 04 09 2017 05 09
2100 02 01 2100 03 01
9996 02 01 9996 03 01
2000 01 01 9999 12 31

1
2017 04 09 2017 05 09

1
2010 04 09 2017 05 09

32
2000 01 01 9999 12 31
2000 01 01 9999 12 31
2000 01 01 9999 12 31
2000 01 01 9999 12 31
2000 01 01 9999 12 31
2000 01 01 9999 12 31
2000 01 01 9999 12 31
2000 01 01 9999 12 31
2000 01 01 9999 12 31
2000 01 01 9999 12 31
2000 01 01 9999 12 31
2000 01 01 9999 12 31
2000 01 01 9999 12 31
2000 01 01 9999 12 31
2000 01 01 9999 12 31
2000 01 01 9999 12 31
2000 01 01 9999 12 31
2000 01 01 9999 12 31
2000 01 01 9999 12 31
2000 01 01 9999 12 31
2000 01 01 9999 12 31
2000 01 01 9999 12 31
2000 01 01 9999 12 31
2000 01 01 9999 12 31
2000 01 01 9999 12 31
2000 01 01 9999 12 31
2000 01 01 9999 12 31
2000 01 01 9999 12 31
2000 01 01 9999 12 31
2000 01 01 9999 12 31
2000 01 01 9999 12 31
2000 01 01 9999 12 31
'''